### PR TITLE
APP-1517 Change assertion library and assure that tests work

### PR DESCRIPTION
**Pull request by Michael-A-Heuer**
## Description

This PR removes `ethereum-waffle` and replaces `hardhat-waffle` by `hardhat-chai-matchers`.
This required to create `GovernanceERC20Mock` because we used a `waffle` mock for the `TokenVoting` tests, which required adapting the respective tests.

**To be merged directly after https://github.com/aragon/core/pull/176 to ensure that our event tests work.**

Task: [APP-1517](https://aragonassociation.atlassian.net/browse/APP-1517)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran all tests with success and extended them if necessary.
- [x] I have updated the ``CHANGELOG.md`` file in the root folder.
- [x] I have tested my code on the test network.

[APP-1517]: https://aragonassociation.atlassian.net/browse/APP-1517?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/contracts/test/voting/addresslist-voting.ts [hunk 1]
```diff
@@ -96,29 +101,19 @@ describe('AddresslistVoting', function () {
...
-      const addr = signers[i].address;
-      addresses.push(addr);
-    }
-    return addresses;
-  }
-
   describe('initialize: ', async () => {
     it('reverts if trying to re-initialize', async () => {
-      await voting.initialize(dao.address, votingSettings, addresslist(0));
+      await voting.initialize(dao.address, votingSettings, await addresses(0));
```
Comment (brickpop) [inline packages/contracts/test/voting/addresslist-voting.ts:106]: Wouldn't it be easier/faster to get the array once at the beginning and keep it stored locally?

File: packages/contracts/test/voting/multisig.ts [hunk 2]
```diff
@@ -338,140 +456,189 @@ describe('Multisig', function () {
...
+
+        await expect(multisig.connect(signers[3]).execute(id))
+          .to.emit(dao, DAO_EVENTS.EXECUTED)
+          .to.emit(multisig, PROPOSAL_EVENTS.PROPOSAL_EXECUTED)
+          .to.not.emit(multisig, MULTISIG_EVENTS.APPROVED);
+      });
+
+      it('emits the `Approved`, `ProposalExecuted`, and `Executed` events if execute is called inside the `approve` method', async () => {
+        await multisig.connect(signers[0]).approve(id, false);
+        await multisig.connect(signers[1]).approve(id, false);
```
Comment (brickpop) [inline packages/contracts/test/voting/multisig.ts:635]: It seems like we are changing code beyond the assertion library itself here?

The logic tested at the end seems unrelated

Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/multisig.ts]: This was because this PR branched off of #176. Now, that is merged, these changes are not present on this branch anymore.

File: packages/contracts/test/voting/addresslist-voting.ts (hunk 1 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/addresslist-voting.ts:106]: You mean in an `beforeEach` block? 
The array length often varies across tests. I would keep it as is

File: packages/contracts/test/voting/token-voting.ts [hunk 3]
```diff
@@ -107,6 +116,28 @@ describe('TokenVoting', function () {
     );
   });
 
+  async function setBalances(
+    receivers: string[],
+    amounts: number[],
+    totalSupply: number
+  ) {
+    expect(receivers.length).to.equal(amounts.length);
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:124]: I personally don't like having `expect` in helper functions. they don't check if a state is right 
also why don't you pass in an array of `{receiver: string, amount: number}`?

File: packages/contracts/test/voting/token-voting.ts [hunk 4]
```diff
@@ -107,6 +116,28 @@ describe('TokenVoting', function () {
...
+
+    expect(summedAmounts).to.be.lte(totalSupply);
+
+    // Mint the difference between `totalSupply` and `summedAmounts` to `address(0)`
+    if (summedAmounts < totalSupply) {
+      receivers.push(`0x${'dead'.repeat(10)}`);
+      amounts.push(totalSupply - summedAmounts);
+    }
+
+    for (let i = 0; i < receivers.length; i++) {
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:136]: run that in a non blocking loop to speed things up:
```typescript
const promises = receivers.map((recv, i)=> governanceErc20Mock.setBalance(recv, amounts[i]))
await Promise.all(promises)
```

File: packages/contracts/test/voting/token-voting.ts [hunk 5]
```diff
@@ -496,34 +521,32 @@ describe('TokenVoting', function () {
...
         await advanceIntoVoteTime(startDate, endDate);
 
-        await governanceErc20Mock.mock.getPastVotes.returns(51);
-        await voting.vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[0]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[1]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[2]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[3]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[4]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[5]).vote(id, VoteOption.Yes, false);
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:549]: could also be optimized with a loop and `Promise.all`

File: packages/contracts/test/voting/token-voting.ts [hunk 6]
```diff
@@ -537,12 +560,13 @@ describe('TokenVoting', function () {
...
-        await governanceErc20Mock.mock.getPastVotes.returns(20);
-        await voting.connect(signers[1]).vote(id, VoteOption.No, false);
-        await governanceErc20Mock.mock.getPastVotes.returns(20);
-        await voting.connect(signers[2]).vote(id, VoteOption.Abstain, false);
+        await voting.connect(signers[1]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[2]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[3]).vote(id, VoteOption.No, false);
+        await voting.connect(signers[4]).vote(id, VoteOption.No, false);
+        await voting.connect(signers[5]).vote(id, VoteOption.Abstain, false);
+        await voting.connect(signers[6]).vote(id, VoteOption.Abstain, false);
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:569]: could also be optimized with a loop and `Promise.all`

File: packages/contracts/test/voting/token-voting.ts [hunk 7]
```diff
@@ -567,17 +591,19 @@ describe('TokenVoting', function () {
...
         await advanceIntoVoteTime(startDate, endDate);
 
-        await governanceErc20Mock.mock.getPastVotes.returns(50);
-        await voting.vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[0]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[1]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[2]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[3]).vote(id, VoteOption.Yes, false);
+
         expect(await voting.canExecute(id)).to.equal(false);
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:599]: could also be optimized with a loop and `Promise.all`

File: packages/contracts/test/voting/token-voting.ts [hunk 8]
```diff
@@ -637,90 +657,83 @@ describe('TokenVoting', function () {
...
       it('can execute early if participation is large enough', async () => {
         await advanceIntoVoteTime(startDate, endDate);
 
-        await governanceErc20Mock.mock.getPastVotes.returns(50);
-        await voting.vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[0]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[1]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[2]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[3]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[4]).vote(id, VoteOption.Yes, false);
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:736]: could also be optimized with a loop and `Promise.all`

File: packages/contracts/test/voting/token-voting.ts [hunk 9]
```diff
@@ -756,16 +767,19 @@ describe('TokenVoting', function () {
...
 
-        // vote 30 voting no votes
-        await governanceErc20Mock.mock.getPastVotes.returns(30);
-        await voting.connect(signers[1]).vote(id, VoteOption.No, false);
+        // vote with 30 no votes
+        await voting.connect(signers[5]).vote(id, VoteOption.No, false);
+        await voting.connect(signers[6]).vote(id, VoteOption.No, false);
+        await voting.connect(signers[7]).vote(id, VoteOption.No, false);
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:779]: could also be optimized with a loop and `Promise.all`

File: packages/contracts/test/voting/token-voting.ts [hunk 10]
```diff
@@ -915,8 +919,12 @@ describe('TokenVoting', function () {
...
         await advanceIntoVoteTime(startDate, endDate);
 
-        await governanceErc20Mock.mock.getPastVotes.returns(51);
-        await voting.vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[0]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[1]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[2]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[3]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[4]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[5]).vote(id, VoteOption.Yes, false);
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:928]: could also be optimized with a loop and `Promise.all`

File: packages/contracts/test/voting/token-voting.ts [hunk 11]
```diff
@@ -930,12 +938,15 @@ describe('TokenVoting', function () {
...
-        await voting.vote(id, VoteOption.Abstain, false);
+        await voting.connect(signers[0]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[1]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[2]).vote(id, VoteOption.Yes, false);
+
+        await voting.connect(signers[3]).vote(id, VoteOption.No, false);
+        await voting.connect(signers[4]).vote(id, VoteOption.No, false);
+
+        await voting.connect(signers[5]).vote(id, VoteOption.Abstain, false);
+        await voting.connect(signers[6]).vote(id, VoteOption.Abstain, false);
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:949]: could also be optimized with a loop and `Promise.all`

File: packages/contracts/test/voting/token-voting.ts [hunk 12]
```diff
@@ -1104,8 +1114,11 @@ describe('TokenVoting', function () {
       });
 
       it('executes early if participation and support are met and the vote outcome cannot change anymore', async () => {
-        await governanceErc20Mock.mock.getPastVotes.returns(50);
         await voting.connect(signers[0]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[1]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[2]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[3]).vote(id, VoteOption.Yes, false);
+        await voting.connect(signers[4]).vote(id, VoteOption.Yes, false);
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:1121]: could also be optimized with a loop and `Promise.all`

**Review by mathewmeconry — CHANGES_REQUESTED** [review]

File: packages/contracts/test/voting/token-voting.ts (hunk 4 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts]: Fixed in 08aa9e5c57d9581a9960d89c3329b258823f8637

File: packages/contracts/test/voting/token-voting.ts (hunk 3 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts]: Fixed in 08aa9e5c57d9581a9960d89c3329b258823f8637

File: packages/contracts/test/voting/token-voting.ts (hunk 9 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts]: Improved in e66e59747b75c6ce535bde30be1342d74a358978

File: packages/contracts/test/voting/token-voting.ts (hunk 8 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts]: Improved in e66e59747b75c6ce535bde30be1342d74a358978

File: packages/contracts/test/voting/token-voting.ts (hunk 7 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts:607]: Improved in e66e59747b75c6ce535bde30be1342d74a358978

File: packages/contracts/test/voting/token-voting.ts (hunk 6 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts]: Improved in e66e59747b75c6ce535bde30be1342d74a358978

File: packages/contracts/test/voting/token-voting.ts (hunk 5 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts]: Improved in e66e59747b75c6ce535bde30be1342d74a358978

File: packages/contracts/test/voting/token-voting.ts (hunk 4 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts]: Improved in e66e59747b75c6ce535bde30be1342d74a358978

File: packages/contracts/test/voting/token-voting.ts (hunk 3 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts]: Improved in bbc468ded7f531d27ce685c55d3dd166acc20bef

File: packages/contracts/test/voting/token-voting.ts (hunk 12 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts]: Improved in bbc468ded7f531d27ce685c55d3dd166acc20bef

File: packages/contracts/test/voting/token-voting.ts (hunk 10 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts:915]: Improved in bbc468ded7f531d27ce685c55d3dd166acc20bef

File: packages/contracts/test/voting/addresslist-voting.ts (hunk 1 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/addresslist-voting.ts]: Removed the `addresses` helper in 0d329872360849b5a4a5a8aa53d88388309e8b4c

File: packages/contracts/test/test-utils/voting.ts [hunk 13]
```diff
@@ -57,3 +58,33 @@ export async function advanceAfterVoteEnd(endDate: number) {
...
+    yes: number[];
+    no: number[];
+    abstain: number[];
+  }
+) {
+  let promises = signerIds.yes.map(i =>
+    votingContract.connect(signers[i]).vote(proposalId, VoteOption.Yes, false)
+  );
+
+  promises.push(
```
Comment (mathewmeconry) [inline packages/contracts/test/test-utils/voting.ts:76]: I guess you wanted to append the array from the no loop the array of the yes loop. otherwise `Promise.all` doesn't work as you would expect.
use `concat` instead of `push`: `promises = promise.concat(....)`

File: packages/contracts/test/voting/addresslist-voting.ts [hunk 14]
```diff
@@ -419,7 +421,11 @@ describe('AddresslistVoting', function () {
       beforeEach(async () => {
         votingSettings.votingMode = VotingMode.Standard;
 
-        await voting.initialize(dao.address, votingSettings, addresslist(10));
+        await voting.initialize(
+          dao.address,
+          votingSettings,
+          [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map(i => signers[i].address)
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/addresslist-voting.ts:427]: using `slice(...)` with `map(...)` would probably be cleaner: `signers.slice(0, 9).map(s => s.address)`

File: packages/contracts/test/voting/addresslist-voting.ts [hunk 15]
```diff
@@ -509,54 +508,54 @@ describe('AddresslistVoting', function () {
...
 
     context('Early Execution Mode', async () => {
       beforeEach(async () => {
         votingSettings.votingMode = VotingMode.EarlyExecution;
 
-        await voting.initialize(dao.address, votingSettings, addresslist(10));
+        await voting.initialize(
+          dao.address,
+          votingSettings,
+          [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map(i => signers[i].address)
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/addresslist-voting.ts:557]: using `slice(...)` with `map(...)` would probably be cleaner: `signers.slice(0, 9).map(s => s.address)`

File: packages/contracts/test/voting/addresslist-voting.ts [hunk 16]
```diff
@@ -736,25 +726,29 @@ describe('AddresslistVoting', function () {
...
 
     context('Vote Replacement Mode', async () => {
       beforeEach(async () => {
         votingSettings.votingMode = VotingMode.VoteReplacement;
 
-        await voting.initialize(dao.address, votingSettings, addresslist(10));
+        await voting.initialize(
+          dao.address,
+          votingSettings,
+          [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map(i => signers[i].address)
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/addresslist-voting.ts:750]: using `slice(...)` with `map(...)` would probably be cleaner: `signers.slice(0, 9).map(s => s.address)`

File: packages/contracts/test/voting/addresslist-voting.ts [hunk 17]
```diff
@@ -897,7 +881,11 @@ describe('AddresslistVoting', function () {
       beforeEach(async () => {
         votingSettings.minParticipation = pct16(25);
 
-        await voting.initialize(dao.address, votingSettings, addresslist(10));
+        await voting.initialize(
+          dao.address,
+          votingSettings,
+          [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map(i => signers[i].address)
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/addresslist-voting.ts:887]: using `slice(...)` with `map(...)` would probably be cleaner: `signers.slice(0, 9).map(s => s.address)`

File: packages/contracts/test/voting/addresslist-voting.ts [hunk 18]
```diff
@@ -1226,7 +1214,11 @@ describe('AddresslistVoting', function () {
         votingSettings.supportThreshold = pct16(0);
         votingSettings.minParticipation = pct16(0);
 
-        await voting.initialize(dao.address, votingSettings, addresslist(10));
+        await voting.initialize(
+          dao.address,
+          votingSettings,
+          [0, 1, 2, 3, 4, 5, 6, 7, 8, 9].map(i => signers[i].address)
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/addresslist-voting.ts:1220]: using `slice(...)` with `map(...)` would probably be cleaner: `signers.slice(0, 9).map(s => s.address)`

File: packages/contracts/test/voting/token-voting.ts [hunk 19]
```diff
@@ -121,11 +152,16 @@ describe('TokenVoting', function () {
...
         )
-      ).to.be.revertedWith(ERRORS.ALREADY_INITIALIZED);
+      ).to.be.revertedWith('Initializable: contract is already initialized');
     });
   });
 
   describe('Proposal creation', async () => {
+    beforeEach(async () => {
+      await setBalances([{receiver: signers[0].address, amount: 1}]);
+      await setTotalSupply(1);
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:162]: unecessary call because the `totalSupply` would already be 1 after `setBalances`

File: packages/contracts/test/voting/token-voting.ts [hunk 20]
```diff
@@ -107,6 +118,26 @@ describe('TokenVoting', function () {
...
   });
 
+  async function setBalances(balances: {receiver: string; amount: number}[]) {
+    const promises = balances.map(balance =>
+      governanceErc20Mock.setBalance(balance.receiver, balance.amount)
+    );
+    await Promise.all(promises);
+  }
+
+  async function setTotalSupply(totalSupply: number) {
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:128]: I think that this function should either throw an error when the passed `totalSupply` is below the current or start burning tokens. 
In tests when I set something I expect it to be that value afterwards

File: packages/contracts/test/voting/token-voting.ts [hunk 21]
```diff
@@ -167,52 +201,69 @@ describe('TokenVoting', function () {
     it('reverts if the user is not allowed to create a proposal and minProposerPower > 1 is selected', async () => {
       votingSettings.minProposerVotingPower = 123;
 
+      await setBalances([
+        {
+          receiver: signers[1].address,
+          amount: votingSettings.minProposerVotingPower,
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:207]: set the amount to `votingSettings.minProposerVotingPower - 1` and safe the call to `setTotalSupply`

File: packages/contracts/test/voting/token-voting.ts [hunk 22]
```diff
@@ -463,6 +489,23 @@ describe('TokenVoting', function () {
   });
 
   describe('Proposal + Execute:', async () => {
+    beforeEach(async () => {
+      const receivers = Array.from(Array(12).keys()).map(
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:493]: either the `slice` option or the proposed helper

File: packages/contracts/test/voting/token-voting.ts [hunk 23]
```diff
@@ -463,6 +489,23 @@ describe('TokenVoting', function () {
...
+
+      const balances = receivers.map((receiver, i) => {
+        return {
+          receiver: receiver,
+          amount: amounts[i],
+        };
+      });
+
+      await setBalances(balances);
+      await setTotalSupply(100);
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/token-voting.ts:506]: totalSupply is already 100, isn't it?
9 * 10 + 5 + 4 + 1 = 100

**Review by mathewmeconry — CHANGES_REQUESTED** [review]
`[0, 1, 2, 3, 4].map(i => signers[i].address),` gets uses quite a lot. Maybe even a helper function would make sende `getSignersAddresses(signers, amount)`
maybe even better than the `slice` solution I proposed

File: packages/contracts/test/test-utils/voting.ts (hunk 13 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/test-utils/voting.ts]: Fixed 53df3eb707971b0343ca6ede46b2205b13e4198d

Comment (Michael-A-Heuer) [reply packages/contracts/test/test-utils/voting.ts]: Thx, strange that it worked

File: packages/contracts/test/voting/token-voting.ts (hunk 23 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts:506]: I wanted to write it explicitly.
Yes, `setTotalSupply` sets the balance of `address(1)` to `BigNumber.from(totalSupply).sub(currentTotalSupply)` so it is not increased in this case. However, I wanted to explicitly set the value to 100 to be sure that this is the total supply.
If I set it to `setTotalSupply(99)`, the tests would fail `Error: value out-of-bounds` which is good.

The mocking functionality should be improved / replaced in a later task. It is not good, admittedly.

File: packages/contracts/test/voting/addresslist-voting.ts (hunk 14 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/addresslist-voting.ts]: Fixed in 26e42329bc5434e96fd75c0bb61139616678b262

File: packages/contracts/test/voting/token-voting.ts (hunk 22 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts]: Fixed in 26e42329bc5434e96fd75c0bb61139616678b262

File: packages/contracts/test/voting/token-voting.ts (hunk 21 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts:207]: Clarified the test in e78f9088d9efe40e038e3f1dc7e6a61ef03c1d53

File: packages/contracts/test/voting/token-voting.ts (hunk 20 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts:128]: It throws `Error: value out-of-bounds (argument="amount", value={"type":"BigNumber","hex":"-0x01"}, code=INVALID_ARGUMENT, version=abi/5.6.0)`

Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts:128]: Should I throw an explicit error?

File: packages/contracts/test/test-utils/voting.ts [hunk 24]
```diff
@@ -73,12 +73,12 @@ export async function voteWithSigners(
     votingContract.connect(signers[i]).vote(proposalId, VoteOption.Yes, false)
   );
 
-  promises.push(
+  promises.concat(
```
Comment (mathewmeconry) [inline packages/contracts/test/test-utils/voting.ts:76]: `promises.concat(...)` returns a new array and doesn't change the exising array. therefore write `promises = promises.concat(...)`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat

**Review by mathewmeconry — CHANGES_REQUESTED** [review]

File: packages/contracts/test/voting/token-voting.ts (hunk 19 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts:162]: Same as here: https://github.com/aragon/core/pull/178#discussion_r1065773796

File: packages/contracts/test/voting/addresslist-voting.ts (hunk 17 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/addresslist-voting.ts]: Fixed here 26e42329bc5434e96fd75c0bb61139616678b262

File: packages/contracts/test/voting/addresslist-voting.ts (hunk 15 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/addresslist-voting.ts]: Fixed here 26e42329bc5434e96fd75c0bb61139616678b262

File: packages/contracts/test/test-utils/voting.ts (hunk 24 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/test-utils/voting.ts]: Fixed in 1db1130b866b84c7ebe605e2ba62ee66674ae504

File: packages/contracts/test/voting/token-voting.ts (hunk 23 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts:502]: I can also write an expect statement

File: packages/contracts/test/voting/addresslist-voting.ts (hunk 18 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/addresslist-voting.ts]: Fixed in 26e42329bc5434e96fd75c0bb61139616678b262

**Review by mathewmeconry — APPROVED** [review]

**Review by novaknole — DISMISSED** [review]

File: packages/contracts/test/core/dao.ts [hunk 25]
```diff
@@ -115,7 +115,7 @@ describe('DAO', function () {
     it('reverts if trying to re-initialize', async () => {
       await expect(
         dao.initialize(dummyMetadata1, ownerAddress, dummyAddress1)
-      ).to.be.revertedWith(ERRORS.ALREADY_INITIALIZED);
+      ).to.be.revertedWith('Initializable: contract is already initialized');
```
[the remark this answers is not in the document]
Comment (Michael-A-Heuer) [reply packages/contracts/test/core/dao.ts]: fa9bc62573fbdf6315839d506fe46ce2ff9b7f24

**Review by mathewmeconry — APPROVED** [review]
